File: python/year_2015/day_16.py
"""
https://adventofcode.com/2015/day/16
"""


from pathlib import Path
import re

TARGET = {
    "children": "3",
    "cats": "7",
    "samoyeds": "2",
    "pomeranians": "3",
    "akitas": "0",
    "vizslas": "0",
    "goldfish": "5",
    "trees": "3",
    "cars": "2",
    "perfumes": "1",
}


def find_sue(text, valid_prop):
    def valid_sue(line):
        return all(map(valid_prop, r.sub("", line).split(", ")))

    r = re.compile(r"Sue (\d+): ")

    sue = next(filter(valid_sue, text.splitlines()))

    return int(r.match(sue).groups()[0])


def part_1(text):
    def valid_prop(record):
        key, value = record.split(": ")
        return TARGET[key] == value

    return find_sue(text, valid_prop)


def part_2(text):
    def valid_prop(record):
        key, value = record.split(": ")
        if key == "cats" or key == "trees":
            return TARGET[key] < value
        elif key == "pomeranians" or key == "goldfish":
            return value < TARGET[key]

        return TARGET[key] == value

    return find_sue(text, valid_prop)


def _read_input() -> str:
    return (Path(__file__).with_name("test_data") / "day_16.txt").read_text()


def test_part_1_input():
    assert part_1(_read_input()) == 213

def test_part_2_input():
    assert part_2(_read_input()) == 323

File: rust/year_2015/src/day_16.rs
pub fn part_1(input: &str) -> usize {
    parse_compounds(input)
        .position(|compounds| {
            matches!(
                compounds,
                Compounds {
                    children: Some(3) | None,
                    cats: Some(7) | None,
                    samoyeds: Some(2) | None,
                    pomeranians: Some(3) | None,
                    akitas: Some(0) | None,
                    vizslas: Some(0) | None,
                    goldfish: Some(5) | None,
                    trees: Some(3) | None,
                    cars: Some(2) | None,
                    perfumes: Some(1) | None,
                }
            )
        })
        .unwrap()
        + 1
}

pub fn part_2(input: &str) -> usize {
    parse_compounds(input)
        .position(|compounds| {
            matches!(
                compounds,
                Compounds {
                    children: Some(3) | None,
                    cats: Some(8..) | None,
                    samoyeds: Some(2) | None,
                    pomeranians: Some(..3) | None,
                    akitas: Some(0) | None,
                    vizslas: Some(0) | None,
                    goldfish: Some(..5) | None,
                    trees: Some(4..) | None,
                    cars: Some(2) | None,
                    perfumes: Some(1) | None,
                }
            )
        })
        .unwrap()
        + 1
}

fn parse_compounds(input: &str) -> impl Iterator<Item = Compounds> {
    input.lines().map(|line| {
        let mut compounds = Compounds::default();
        let (_sue, parts) = line.split_once(": ").unwrap();
        for part in parts.split(", ") {
            let (name, value) = part.split_once(": ").unwrap();
            let entry = match name {
                "children" => &mut compounds.children,
                "cats" => &mut compounds.cats,
                "samoyeds" => &mut compounds.samoyeds,
                "pomeranians" => &mut compounds.pomeranians,
                "akitas" => &mut compounds.akitas,
                "vizslas" => &mut compounds.vizslas,
                "goldfish" => &mut compounds.goldfish,
                "trees" => &mut compounds.trees,
                "cars" => &mut compounds.cars,
                "perfumes" => &mut compounds.perfumes,
                _ => panic!("unknown compound {name}"),
            };
            *entry = Some(value.parse().unwrap());
        }
        compounds
    })
}

#[derive(Default)]
struct Compounds {
    children: Option<usize>,
    cats: Option<usize>,
    samoyeds: Option<usize>,
    pomeranians: Option<usize>,
    akitas: Option<usize>,
    vizslas: Option<usize>,
    goldfish: Option<usize>,
    trees: Option<usize>,
    cars: Option<usize>,
    perfumes: Option<usize>,
}

#[cfg(test)]
mod test {
    use test_case::test_case;

    const INPUT: &str = include_str!("../test_data/day_16.txt");

    #[test_case(INPUT => 213)]
    fn part_1(input: &str) -> usize {
        super::part_1(input)
    }

    #[test_case(INPUT => 323)]
    fn part_2(input: &str) -> usize {
        super::part_2(input)
    }
}
